colcon_core/package_identification/python.py

```diff
@@ -97,8 +98,70 @@ def extract_dependencies(options):
     for option_name, dependency_type in mapping.items():
         dependencies[dependency_type] = set()
         for dep in options.get(option_name, []):
-            # remove environmental markers (separated by semicolons)
-            # and version specifiers (separated by comparison operators)
-            name = re.split(r';|<|>|<=|>=|==|!=', dep)[0].rstrip()
-            dependencies[dependency_type].add(name)
+            dependencies[dependency_type].add(
+                create_dependency_descriptor(dep))
     return dependencies
+
+
+def create_dependency_descriptor(requirement_string):
+    """
+    Create a DependencyDescriptor from a PEP440 compliant string.
+
+    See https://www.python.org/dev/peps/pep-0440/#version-specifiers
+
+    :param str requirement_string: a PEP440 compliant requirement string
+    :return: A descriptor with version constraints from the requirement string
+    :rtype: DependencyDescriptor
+    """
+    symbol_mapping = {
+        '==': 'version_eq',
+        '!=': 'version_neq',
+        '<=': 'version_lte',
+        '>=': 'version_gte',
+        '>': 'version_gt',
+        '<': 'version_lt',
+    }
+
+    requirement = parse_requirement(requirement_string)
+    metadata = {}
+    for symbol, version in (requirement.constraints or []):
+        if symbol in symbol_mapping:
+            metadata[symbol_mapping[symbol]] = version
+        elif symbol == '~=':
+            metadata['version_gte'] = version
+            metadata['version_lt'] = _next_incompatible_version(version)
+        else:
+            logger.warn(
+                "Ignoring unknown symbol '{symbol}' in '{requirement}'"
+                .format(locals()))
+    return DependencyDescriptor(requirement.name, metadata=metadata)
+
+
+def _next_incompatible_version(version):
+    """
+    Find the next non-compatible version.
+
+    This is for use with the ~= compatible syntax. It will provide
+    the first version that this version must be less than in order
+    to be compatible.
+
+    :param str version: PEP 440 compliant version number
+    :return: The first version after this version that is not compatible
+    :rtype: str
+    """
+    normalized = NormalizedVersion(version)
+    parse_tuple = normalized.parse(version)
+    version_tuple = parse_tuple[1]
+    limit = len(version_tuple) - 1
+    lt_string = ''
+    for i in range(limit):
+        if i != 0:
+            lt_string += '.'
+        v = version_tuple[i]
+        if i == limit - 1:
+            v += 1
+        lt_string += str(v)
+    if len(lt_string) == 1:
```

Comment:
This condition fails for major versions greater than 9. I will update the test and then refactor the code...